Improve Kotlin tour (#4321)
JetBrains/kotlin-web-site: docs/topics/tour/kotlin-tour-control-flow.md

Reviewer: This section is much longer than the rest. Maybe the conditionals and the loops can go in different pages?

Author: As agreed, I moved some exercises after the ranges section to try and break up the content a bit.

Kotlin provides `if` and `when` for checking conditional expressions. 

> If you have to choose between `if` and `when`, we recommend using `when` because it:

> > * makes your code easier to read.

Reviewer 2: Inconsistent capitalization in lists

> * makes it easier to add another branch.

> > * leads to more robust and safer programs.

Author: Updated to "fewer mistakes" like we discussed.

### When

Use `when` when you have a conditional expression with multiple branches.

To use `when`:
* Place the value you want to evaluate within parentheses `()`.
* Place the branches within curly braces `{}`.
* Use `->` in each branch to separate each check from the action to take if the check is successful.

`when` can be used either as a statement or as an expression.

> A **statement** doesn't return anything but performs actions instead.

Author: Good idea! Fixed!

### When

Use `when` when you have a conditional expression with multiple branches.

To use `when`:
* Place the conditional expression within parentheses `()` and the actions to take
  within curly braces `{}`.

> * Use `->` in each branch to separate each condition from each action.

Reviewer: Properly speaking, the thing between parentheses is not a conditional expression, since you can write anything there, and it will be tested against any value in the branches.

```suggestion
To use `when`:
* Place the value you want to test against within parentheses `()` and the branches
  within curly braces `{}`.
* Use `->` in each branch to separate each test from the action to take in that case.
```

Author: Fixed!

> Use `when` when you have a conditional expression with multiple branches.
> `when` can be used either as a statement or as an expression.

Reviewer: Maybe "refactor" the common parts here?

```suggestion
Use `when` when you have a conditional expression with multiple branches.
* Place the conditional expression within parentheses `()` and the actions to take
within curly braces `{}`. 
* Use `->` in each branch to separate each condition from each action.
Furthermore, `when` can be used either as a statement or as an expression.
```

Author: Section reworked!

A **statement** doesn't return anything but performs actions instead.

Here is an example of using `when` as a statement:

> * Place the conditional expression within parentheses `()` and the actions to take

Reviewer: This sentence may give the wrong impression that it only applies to the case in which `when` is used as a statement (even though you later mention the same syntax for using as an expression).

Author: Section reworked!

>
{type="note"}

> An **expression** returns a value that can be used later on in your code.

Reviewer 2: ```suggestion
An **expression** returns a value that can be used later in your code.
```

* `'a'..'d'` is equivalent to `'a', 'b', 'c', 'd'`
* `'z' downTo 's' step 2` is equivalent to `'z', 'x', 'v', 't'`

## Conditional expressions and ranges practice

### Exercise 1 {initial-collapse-state="collapsed" id="conditional-expressions-exercise-1"}

Create a simple game where you win if throwing two dice results in the same number. Use `if` to print `You win :)`
if the dice match or `You lose :(` otherwise.

<deflist collapsible="true">
    <def title="Hint">

>         Use the <a href="operator-overloading.md#equality-and-inequality-operators"> equality operator </a>  (<code>==</code>) to compare the dice results. 

Reviewer 2: ```suggestion
        Use the <a href="operator-overloading.md#equality-and-inequality-operators">equality operator</a> (<code>==</code>) to compare the dice results. 
```

    </def>
</deflist>

> |---|---|

Reviewer 2: I'm a bit confused by this syntax, I don't see it changing anything on staging
Do two code samples actually go into columns in a table
Same for exercise 2

Author: It's a way of introducing a line break so that the formatting looks nicer.

```kotlin
fun main() {
    val firstResult = Random.nextInt(6)
    val secondResult = Random.nextInt(6)
    // Write your code here
}
```
{validate="false" kotlin-runnable="true" kotlin-min-compiler-version="1.3" id="kotlin-tour-control-flow-conditional-exercise-1"}

|---|---|
```kotlin
fun main() {
    val firstResult = Random.nextInt(6)
    val secondResult = Random.nextInt(6)
    if (firstResult == secondResult)
        println("You win :)")
    else
        println("You lose :(")
}
```
{initial-collapse-state="collapsed" collapsed-title="Example solution" id="kotlin-tour-control-flow-conditional-solution-1"}

### Exercise 2 {initial-collapse-state="collapsed" id="conditional-expressions-exercise-2"}

Using a `when` expression, update the following program so that when you input the names of GameBoy buttons, the actions

> are printed to output.

Reviewer 2: ```suggestion
are printed to the output.
```

Create a simple game where you win if throwing two dice results in the same number. Use `if` to print `You win :)`
if the dice match or `You lose :(` otherwise.

<deflist collapsible="true">
    <def title="Hint">
        Use the <a href="operator-overloading.md#equality-and-inequality-operators"> equality operator </a>  (<code>==</code>) to compare the dice results. 
    </def>
</deflist>

|---|---|
```kotlin
fun main() {
    val firstResult = Random.nextInt(6)
    val secondResult = Random.nextInt(6)
    // Write your code here
}
```
{validate="false" kotlin-runnable="true" kotlin-min-compiler-version="1.3" id="kotlin-tour-control-flow-conditional-exercise-1"}

|---|---|
```kotlin
fun main() {
    val firstResult = Random.nextInt(6)
    val secondResult = Random.nextInt(6)
    if (firstResult == secondResult)
        println("You win :)")
    else
        println("You lose :(")
}
```
{initial-collapse-state="collapsed" collapsed-title="Example solution" id="kotlin-tour-control-flow-conditional-solution-1"}

### Exercise 2 {initial-collapse-state="collapsed" id="conditional-expressions-exercise-2"}

> Using a `when` expression, update the following program so that when you input the names of GameBoy buttons, the actions

Reviewer: You mean "X = jump", "Y = shoot fire", as in Mario Bros, right? :P

Author: No. This is for pokemon! :D

    </def>
</deflist>

<deflist collapsible="true">

>     <def title="Hint">

Reviewer 2: ```suggestion
    <def title="Hint 2">
```

    }
}
```
{initial-collapse-state="collapsed" collapsed-title="Example solution" id="kotlin-tour-control-flow-loops-solution-3"}

> 

Reviewer 2: ```suggestion
```
